executors.md

```diff
@@ -0,0 +1,179 @@
+  multithreaded execution.
+
+It brings the futures library into line with the design principles of the [Tokio
+Reform], but also makes some improvements to the `current_thread` design.
+
+# Proposed design
+
+## The core executor abstraction: `Spawn`
+
+First, in `futures-core` we provide a basic definition of a general-purpose
+executor:
+
+```rust
+trait Spawn {
+    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>);
+}
+```
+
+The `Spawn` trait is quite straightforward (some alternatives and tradeoffs are
+discussed in the next section), and crucially is object-safe.
+
+We then build in an executor to the task context, stored internally as a trait
+object, which allows us to provide the following methods:
+
+```rust
+impl task::Context {
+    // Spawn onto the current default executor
+    fn spawn<F>(&self, F)
+        where F: Future<Item = (), Error = ()> + Send + 'static;
+
+    // Run the given closure with a context using a new default executor
+    fn with_spawn<S, F, R>(&mut self, spawn: S, f: F) -> R
+        where S: Spawn, F: FnOnce(&mut task::Context) -> R;
+}
+```
+
+With those APIs in place, we've achieved two of our goals: it's possible for any
+future to spawn a new task, and to exert fine-grained control over generic task
+spawning within a sub-future.
+
+## Built-in executors
+
+In the `futures-executor` crate, we then add two executors: `ThreadPool` and
+`LocalPool`, providing multi-threaded and single-threaded execution,
+respectively.
+
+### Multi-threaded execution: `ThreadPool`
+
+The `ThreadPool` executor works basically like `CpuPool` today:
+
+```rust
+struct ThreadPool { ... }
+impl Spawn for ThreadPool { ... }
+
+impl ThreadPool {
+    // sets up a pool with the default number of threads
+    fn new() -> ThreadPool;
+
+    // starts up the worker threads
+    fn run();
+}
+```
+
+Tasks spawn onto a `ThreadPool` will, by default, spawn any subtasks onto the
+same executor.
+
+### Single-threaded execution: `LocalPool`
+
+The `LocalPool` API, on the other hand, is a bit more subtle. Here, we're
+replacing `current_thread` from [Tokio Reform] with a slightly different, more
+flexible design that integrates with the default executor system.
+
+First, we have the basic type definition and executor definition, which is much
+like `ThreadPool`:
+
+```rust
+// Note: not `Send` or `Sync`
+struct LocalPool { ... }
+impl Spawn for LocalPool { .. }
+
+impl LocalPool {
+    // create a new single-threaded executor
+    fn new() -> LocalPool;
+}
+```
+
+However, the rest of the API is more interesting:
+
+```rust
+impl LocalPool {
+    // runs the executor until `f` is resolved, spawning subtasks onto `spawn`
+    fn run_until<F, S>(&self, f: F, spawn: S) -> Result<F::Item, F::Error>
+        where F: Future, S: Spawn;
+
+    // a future that resolves when *all* spawned futures have resolved
+    fn all_done(&self) -> impl Future<Item = (), Error = ()>;
```

Comment:
IMO it's a little bit confusing right now to figure out how to actually run `all_done`-- it almost looks like you'd need a separate `LocalPool` for it. The trick is that you can do `pool.run_until(pool.all_done())`. It might be nice to offer a `run_until_all_done()` method that offers this same functionality in a more visible way.

Reply:
Yeah, I think this design might be a *bit* too clever for its own good.

Reply:
I'm probably being pedantic, but technically, it should be:

```rust
let pool = LocalPool::new();
pool.spawn_local(...); // start some future
pool.run_until(pool.all_done(), pool); // this doesn't seem possible, since spawn is by value
```

Reply:
@ahmedcharles I'd imagine there'd be an impl of `Spawn` for `&LocalPool` as @carllerche suggested above, which would let you do `pool.run_until(pool.all_done(), &pool);`. Still, that seems pretty unergonomic to me-- I think it'd be good to offer a version of `run_until` which uses `&self` as the spawner, so that you could write `pool.run_until(my_fut);` rather than `pool.run_until(my_fut, &pool);`

Reply:
I see. Makes sense. I'm fine with the current design, since it's easy to add a function which does the common thing later.